fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java

```diff
@@ -277,12 +346,172 @@ public void addEnforcerPlan(GroupExpression groupExpression, Group group) {
+            LogicalProperties logicalProperties) {
+        if (fromGroup == targetGroup) {
+            return;
+        }
+        // simple check targetGroup is the ancestors of the fromGroup, not check completely because of performance
+        if (fromGroup == root) {
+            throw new IllegalStateException(
+                    "TargetGroup should be ancestors of fromGroup, but fromGroup is root. Maybe a bug");
+        }
+
+        List<GroupExpression> logicalExpressions = fromGroup.clearLogicalExpressions();
+        recycleGroup(fromGroup);
+
+        recycleLogicalExpressions(targetGroup);
+        recyclePhysicalExpressions(targetGroup);
+
+        for (GroupExpression logicalExpression : logicalExpressions) {
+            targetGroup.addLogicalExpression(logicalExpression);
+        }
+        targetGroup.setLogicalProperties(logicalProperties);
+    }
+
+    private void reInitGroup(Group group, GroupExpression initLogicalExpression, LogicalProperties logicalProperties) {
+        recycleLogicalExpressions(group);
+        recyclePhysicalExpressions(group);
+
+        group.setLogicalProperties(logicalProperties);
+        group.addLogicalExpression(initLogicalExpression);
+
+        // note: put newGroupExpression must behind recycle existedExpression(reInitGroup method),
+        //       because existedExpression maybe equal to the newGroupExpression and recycle
+        //       existedExpression will recycle newGroupExpression
+        groupExpressions.put(initLogicalExpression, initLogicalExpression);
+    }
+
     private Plan replaceChildrenToGroupPlan(Plan plan, List<Group> childrenGroups) {
+        if (childrenGroups.isEmpty()) {
+            return plan;
+        }
         List<Plan> groupPlanChildren = childrenGroups.stream()
                 .map(GroupPlan::new)
                 .collect(ImmutableList.toImmutableList());
         LogicalProperties logicalProperties = plan.getLogicalProperties();
         return plan.withChildren(groupPlanChildren)
             .withLogicalProperties(Optional.of(logicalProperties));
     }
+
+    private void moveParentExpressionsReference(Group fromGroup, Group toGroup) {
+        for (GroupExpression parentGroupExpression : fromGroup.getParentGroupExpressions()) {
+            /*
+             * the scenarios that 'parentGroupExpression == toGroup': eliminate the root group.
+             * the fromGroup is group 1, the toGroup is group 2, we can not replace group2's
+             * groupExpressions reference the child group which is group 2 (reference itself).
+             *
+             *   A(group 2)            B(group 2)
+             *   |                     |
+             *   B(group 1)      =>    C(group 0)
+             *   |
+             *   C(group 0)
+             */
+            if (parentGroupExpression.getOwnerGroup() != toGroup) {
+                parentGroupExpression.replaceChild(fromGroup, toGroup);
```

Comment:
since group expression's equals and hashCode include children, if we update children list, we need to re-add it to groupExpressions map.

Reply:
the groupExpressions will be put into the map after invoke `reInitGroup`

```java
moveParentExpressionsReference(existedExpression.getOwnerGroup(), targetGroup);
recycleGroup(existedExpression.getOwnerGroup());
// re-add it here
reInitGroup(targetGroup, newExpression, existedPlan.getLogicalProperties());
return CopyInResult.of(true, newExpression);
```

Reply:
if we do this separately, we need to add more comment to explain it. Otherwise this method may be used by mistake